{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dafe2c14-f37c-4849-8fd4-ab09c77817b7",
   "metadata": {
    "collapsed": false,
    "gradient": {
     "editing": false,
     "source_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "!pip install -q gradio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96ed555-a793-44c7-b737-41cd91b47cde",
   "metadata": {
    "collapsed": false,
    "gradient": {
     "editing": false,
     "source_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running locally at: http://127.0.0.1:7867/\n",
      "This share link will expire in 24 hours. If you need a permanent link, visit: https://gradio.app/introducing-hosted (NEW!)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running on External URL: https://21927.gradio.app\n",
      "Interface loading below...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"900\"\n",
       "            height=\"500\"\n",
       "            src=\"https://21927.gradio.app\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f50be7649a0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data",
     "transient": {}
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:03,835] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 333, in process\n",
      "    predictions, durations = self.run_prediction(processed_input, return_duration=True)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 306, in run_prediction\n",
      "    prediction = predict_fn(*processed_input)\n",
      "TypeError: runQuery() takes 6 positional arguments but 7 were given\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:30,584] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 333, in process\n",
      "    predictions, durations = self.run_prediction(processed_input, return_duration=True)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 306, in run_prediction\n",
      "    prediction = predict_fn(*processed_input)\n",
      "TypeError: runQuery() takes 6 positional arguments but 7 were given\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:32,525] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 333, in process\n",
      "    predictions, durations = self.run_prediction(processed_input, return_duration=True)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 306, in run_prediction\n",
      "    prediction = predict_fn(*processed_input)\n",
      "TypeError: runQuery() takes 6 positional arguments but 7 were given\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:33,757] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 333, in process\n",
      "    predictions, durations = self.run_prediction(processed_input, return_duration=True)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 306, in run_prediction\n",
      "    prediction = predict_fn(*processed_input)\n",
      "TypeError: runQuery() takes 6 positional arguments but 7 were given\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:35,059] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 333, in process\n",
      "    predictions, durations = self.run_prediction(processed_input, return_duration=True)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 306, in run_prediction\n",
      "    prediction = predict_fn(*processed_input)\n",
      "TypeError: runQuery() takes 6 positional arguments but 7 were given\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:37,002] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 333, in process\n",
      "    predictions, durations = self.run_prediction(processed_input, return_duration=True)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 306, in run_prediction\n",
      "    prediction = predict_fn(*processed_input)\n",
      "TypeError: runQuery() takes 6 positional arguments but 7 were given\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:40,995] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 333, in process\n",
      "    predictions, durations = self.run_prediction(processed_input, return_duration=True)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 306, in run_prediction\n",
      "    prediction = predict_fn(*processed_input)\n",
      "TypeError: runQuery() takes 6 positional arguments but 7 were given\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-09-02 17:24:42,244] ERROR in app: Exception on /api/predict/ [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask_cors/extension.py\", line 165, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/flask/app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 92, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/networking.py\", line 180, in predict\n",
      "    prediction, durations = app.interface.process(raw_input)\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 331, in process\n",
      "    processed_input = [input_component.preprocess(raw_input[i])\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/interface.py\", line 331, in <listcomp>\n",
      "    processed_input = [input_component.preprocess(raw_input[i])\n",
      "  File \"/opt/conda/lib/python3.8/site-packages/gradio/inputs.py\", line 247, in preprocess\n",
      "    return float(x)\n",
      "ValueError: could not convert string to float: ''\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "\n",
    "import gradio as gr\n",
    "\n",
    "from encoder import *\n",
    "from decoder import *\n",
    "\n",
    "MODELS_PATH = '../models/'\n",
    "availableModels = [f for f in glob.glob(MODELS_PATH + \"*.pt\", recursive=True)]\n",
    "\n",
    "RESOLUTION = 36\n",
    "KEYBOARD_SIZE = 88\n",
    "MIDI_OFFSET = 20\n",
    "\n",
    "\n",
    "def getFileLen(contextFile):\n",
    "  if isinstance(contextFile, gr.inputs.File):\n",
    "    contextFileLen = 2\n",
    "  if isinstance(contextFile, pd.DataFrame):\n",
    "    contextFileLen = len(contextFile)\n",
    "  else:\n",
    "    contextFileLen = len(encode_data( os.path.abspath(contextFile),\n",
    "                                      n_frames=RESOLUTION,\n",
    "                                      n_notes=KEYBOARD_SIZE,\n",
    "                                      midi_offset=MIDI_OFFSET\n",
    "                                ))\n",
    "    return contextFileLen\n",
    "\n",
    "def runQuery( modelPath,\n",
    "              contextFile,\n",
    "              firstContextMeasure,\n",
    "              lastContextMeasure,\n",
    "              amountToGenerate,\n",
    "              temperature):\n",
    "  \n",
    "  model = torch.load(modelPath)\n",
    "  \n",
    "  \n",
    "  # mid -> pkl\n",
    "  encoded_context = encode_data(contextFile,\n",
    "                                n_frames=RESOLUTION,\n",
    "                                n_notes=KEYBOARD_SIZE,\n",
    "                                midi_offset=MIDI_OFFSET)\n",
    "\n",
    "  # Filter by instrument\n",
    "  encoded_context = encoded_context[encoded_context.index==instrument]\n",
    "\n",
    "  # If the file does not contain the instrument\n",
    "  if len(encoded_context) == 0:\n",
    "      raise ValueError(f'Oops, it seems that the selected context file does not contain any ({INSTRUMENT}).')\n",
    "  \n",
    "  # Calclulate frames\n",
    "  first_context_frame = (firstContextMeasure-1)*RESOLUTION\n",
    "  last_context_frame = first_context_frame + (lastContextMeasure-1)*RESOLUTION\n",
    "  first_generated_frame = last_context_frame\n",
    "  last_generated_frame = first_generated_frame + amountToGenerate*RESOLUTION\n",
    "\n",
    "  # Separate Info from stackframe\n",
    "  context_info = getInfo(encoded_context, first_context_frame, last_context_frame)\n",
    "  context_sf = getStackframe(encoded_context, first_context_frame, last_context_frame, to_float=True)\n",
    "  generated_info = getInfo(encoded_context, first_generated_frame, last_generated_frame)\n",
    "\n",
    "  generated_sf = generateMeasures(model,\n",
    "                                  context_sf,\n",
    "                                  RESOLUTION,\n",
    "                                  amount=amountToGenerate,\n",
    "                                  temperature=temperature\n",
    "                                  )\n",
    "\n",
    "  generated_representation = mergeData(generated_info, generated_sf, instrument, MIDI_OFFSET)\n",
    "\n",
    "  generatedFile = decode_data(generated_representation, RESOLUTION)\n",
    "\n",
    "  return generatedFile\n",
    "\n",
    "\n",
    "\n",
    "# INPUTS  \n",
    "\n",
    "modelPath = gr.inputs.Dropdown(availableModels, type=\"value\", default=None, label='Select model')\n",
    "instrument = gr.inputs.Textbox(lines=1, placeholder=None, default=\"\", numeric=False, type=\"str\", label=\"Enter model's intrument\")\n",
    "\n",
    "contextFile = gr.inputs.File(file_count=\"single\", type=\"file\", label='Upload MIDI context file', keep_filename=True)\n",
    "firstContextMeasure = gr.inputs.Slider( minimum=1,\n",
    "                                        maximum=getFileLen(contextFile),\n",
    "                                        default=1,\n",
    "                                        step=1,\n",
    "                                        label='Select the first measure to be fed to the model')\n",
    "lastContextMeasure = gr.inputs.Slider(minimum=1,\n",
    "                                      maximum=getFileLen(contextFile),\n",
    "                                      default=getFileLen(contextFile),\n",
    "                                      step=1,\n",
    "                                      label='Select the last measure to be fed to the model')\n",
    "\n",
    "amountToGenerate = gr.inputs.Number(default=1, label='Select amount of measures to generate')\n",
    "temperature = gr.inputs.Number(default=0.5, label='Select generation temperature (must be between 0 and 1)')\n",
    "\n",
    "\n",
    "# OUTPUTS\n",
    "generatedFile = gr.outputs.File(label='Generated result')\n",
    "\n",
    "\n",
    "iface = gr.Interface( fn=runQuery,\n",
    "                      inputs=[modelPath,\n",
    "                              instrument,\n",
    "                              contextFile,\n",
    "                              firstContextMeasure,\n",
    "                              lastContextMeasure,\n",
    "                              amountToGenerate,\n",
    "                              temperature\n",
    "                      ],\n",
    "                      outputs=generatedFile,\n",
    "                      # live=True,\n",
    "                      title='Bidirectional-LSTM based MIDI generation'\n",
    "                      )\n",
    "iface.launch(share=True, debug=True, inbrowser=True)"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
